File: Midterm Lab Work 3/MP12FormattedTable.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/*
 * Student Name: CANTAROS, RICH ANDREI
 * Machine Problem: MP12 - Display dataset in formatted table output
 * Language: Java
 *
 * Description:
 * This program asks the user for the CSV dataset file path first.
 * It reads and parses the CSV file, then stores valid rows in an ArrayList.
 * After loading the dataset, it displays the records in a clean formatted table.
 * Blank unnamed columns and the placeholder Column1 are ignored for readability.
 */
public class MP12FormattedTable {

    static class CSVData {
        ArrayList<String> headers = new ArrayList<>();
        ArrayList<String[]> rows = new ArrayList<>();
    }

    public static void main(String[] args) {
        java.util.Scanner scanner = new java.util.Scanner(System.in);

        System.out.println("Student: CANTAROS, RICH ANDREI");
        System.out.println("MP12 - Display Dataset in Formatted Table Output");
        System.out.print("Enter CSV dataset file path: ");
        String filePath = scanner.nextLine();

        try {
            CSVData data = loadCSV(filePath);

            if (data.rows.isEmpty()) {
                System.out.println("No valid data rows were found in the dataset.");
                return;
            }

            ArrayList<Integer> displayColumns = getDisplayColumns(data.headers);

            if (displayColumns.isEmpty()) {
                System.out.println("No displayable columns were found.");
                return;
            }

            int[] widths = computeColumnWidths(data, displayColumns);
            int rowNumberWidth = Math.max(4, String.valueOf(data.rows.size()).length() + 2);

            System.out.println("\n========== FORMATTED DATASET TABLE ==========");
            System.out.println("Total Valid Rows: " + data.rows.size());
            System.out.println("Displayed Columns: meaningful columns only");
            System.out.println("=============================================\n");

            printSeparator(displayColumns, widths, rowNumberWidth, data.headers);
            System.out.print(formatCell("No.", rowNumberWidth) + " | ");
            for (int i = 0; i < displayColumns.size(); i++) {
                int colIndex = displayColumns.get(i);
                System.out.print(formatCell(data.headers.get(colIndex), widths[i]));
                if (i < displayColumns.size() - 1) {
                    System.out.print(" | ");
                }
            }
            System.out.println();
            printSeparator(displayColumns, widths, rowNumberWidth, data.headers);

            for (int i = 0; i < data.rows.size(); i++) {
                String[] row = data.rows.get(i);

                System.out.print(formatCell(String.valueOf(i + 1), rowNumberWidth) + " | ");

                for (int j = 0; j < displayColumns.size(); j++) {
                    int colIndex = displayColumns.get(j);
                    String value = safeGet(row, colIndex);
                    System.out.print(formatCell(value, widths[j]));

                    if (j < displayColumns.size() - 1) {
                        System.out.print(" | ");
                    }
                }
                System.out.println();
            }

            printSeparator(displayColumns, widths, rowNumberWidth, data.headers);

        } catch (IOException e) {
            System.out.println("Error: " + e.getMessage());
        } catch (Exception e) {
            System.out.println("Unexpected error: " + e.getMessage());
        }
    }

    public static CSVData loadCSV(String filePath) throws IOException {
        CSVData data = new CSVData();
        BufferedReader reader = new BufferedReader(new FileReader(filePath));

        String line;
        boolean headerFound = false;
        int headerSize = 0;

        while ((line = reader.readLine()) != null) {
            line = line.replace("\uFEFF", "");
            List<String> parsedLine = parseCSVLine(line);

            if (!headerFound) {
                if (isRealHeader(parsedLine)) {
                    data.headers.addAll(parsedLine);
                    headerFound = true;
                    headerSize = parsedLine.size();
                }
                continue;
            }

            while (parsedLine.size() < headerSize) {
                parsedLine.add("");
            }

            if (isRowEmpty(parsedLine)) {
                continue;
            }

            data.rows.add(parsedLine.toArray(new String[0]));
        }

        reader.close();

        if (!headerFound) {
            throw new IOException("CSV header row was not found.");
        }

        return data;
    }

    public static List<String> parseCSVLine(String line) {
        ArrayList<String> fields = new ArrayList<>();
        StringBuilder currentField = new StringBuilder();
        boolean inQuotes = false;

        for (int i = 0; i < line.length(); i++) {
            char ch = line.charAt(i);

            if (ch == '"') {
                if (inQuotes && i + 1 < line.length() && line.charAt(i + 1) == '"') {
                    currentField.append('"');
                    i++;
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (ch == ',' && !inQuotes) {
                fields.add(currentField.toString());
                currentField.setLength(0);
            } else {
                currentField.append(ch);
            }
        }

        fields.add(currentField.toString());
        return fields;
    }

    public static boolean isRealHeader(List<String> fields) {
        if (fields.size() < 8) {
            return false;
        }

        return fields.get(0).trim().equalsIgnoreCase("Candidate")
                && fields.get(1).trim().equalsIgnoreCase("Student/ Faculty/ NTE")
                && fields.get(3).trim().equalsIgnoreCase("Exam")
                && fields.get(6).trim().equalsIgnoreCase("Score");
    }

    public static boolean isRowEmpty(List<String> row) {
        for (String field : row) {
            if (!field.trim().isEmpty()) {
                return false;
            }
        }
        return true;
    }

    public static String safeGet(String[] row, int index) {
        if (index >= 0 && index < row.length) {
            return row[index];
        }
        return "";
    }

    /*
     * getDisplayColumns function:
     * - Keeps only meaningful headers
     * - Ignores blank unnamed columns and the placeholder Column1
     */
    public static ArrayList<Integer> getDisplayColumns(ArrayList<String> headers) {
        ArrayList<Integer> displayColumns = new ArrayList<>();

        for (int i = 0; i < headers.size(); i++) {
            String header = headers.get(i).trim();
            if (!header.isEmpty() && !header.equalsIgnoreCase("Column1")) {
                displayColumns.add(i);
            }
        }

        return displayColumns;
    }

    /*
     * computeColumnWidths function:
     * - Computes the table width of every displayed column
     * - Width is limited to keep the table readable
     */
    public static int[] computeColumnWidths(CSVData data, ArrayList<Integer> displayColumns) {
        int[] widths = new int[displayColumns.size()];

        for (int i = 0; i < displayColumns.size(); i++) {
            int colIndex = displayColumns.get(i);
            int maxWidth = data.headers.get(colIndex).length();

            for (String[] row : data.rows) {
                String value = safeGet(row, colIndex);
                if (value.length() > maxWidth) {
                    maxWidth = value.length();
                }
            }

            widths[i] = Math.min(maxWidth, 35);
        }

        return widths;
    }

    /*
     * formatCell function:
     * - Truncates long text if needed
     * - Pads text so columns align properly
     */
    public static String formatCell(String text, int width) {
        if (text == null) {
            text = "";
        }

        if (text.length() > width) {
            if (width > 3) {
                text = text.substring(0, width - 3) + "...";
            } else {
                text = text.substring(0, width);
            }
        }

        return String.format("%-" + width + "s", text);
    }

    public static void printSeparator(ArrayList<Integer> displayColumns, int[] widths, int rowNumberWidth, ArrayList<String> headers) {
        System.out.print(repeat("-", rowNumberWidth) + "-+-");
        for (int i = 0; i < displayColumns.size(); i++) {
            System.out.print(repeat("-", widths[i]));
            if (i < displayColumns.size() - 1) {
                System.out.print("-+-");
            }
        }
        System.out.println();
    }

    public static String repeat(String text, int count) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < count; i++) {
            builder.append(text);
        }
        return builder.toString();
    }
}

File: Midterm Lab Work 3/mp12_formatted_table.js
/*
 * Student Name: CANTAROS, RICH ANDREI
 * Machine Problem: MP12 - Display dataset in formatted table output
 * Language: JavaScript (Node.js)
 *
 * Description:
 * This program asks the user for the CSV dataset file path first.
 * It reads the file using Node.js, parses the CSV data, and stores valid rows.
 * It then displays the dataset in a clean formatted table.
 * Blank unnamed columns and the placeholder Column1 are ignored for readability.
 */

const fs = require("fs");
const readline = require("readline");

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function askQuestion(question) {
    return new Promise(resolve => rl.question(question, resolve));
}

function parseCSVLine(line) {
    const fields = [];
    let currentField = "";
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const ch = line[i];

        if (ch === '"') {
            if (inQuotes && i + 1 < line.length && line[i + 1] === '"') {
                currentField += '"';
                i++;
            } else {
                inQuotes = !inQuotes;
            }
        } else if (ch === "," && !inQuotes) {
            fields.push(currentField);
            currentField = "";
        } else {
            currentField += ch;
        }
    }

    fields.push(currentField);
    return fields;
}

function isRealHeader(fields) {
    if (fields.length < 8) return false;

    return fields[0].trim().toLowerCase() === "candidate" &&
           fields[1].trim().toLowerCase() === "student/ faculty/ nte" &&
           fields[3].trim().toLowerCase() === "exam" &&
           fields[6].trim().toLowerCase() === "score";
}

function isRowEmpty(row) {
    return row.every(field => field.trim() === "");
}

function safeGet(row, index) {
    return index >= 0 && index < row.length ? row[index] : "";
}

function loadCSV(filePath) {
    const content = fs.readFileSync(filePath, "utf8").replace(/\uFEFF/g, "");
    const lines = content.split(/\r?\n/);

    const data = {
        headers: [],
        rows: []
    };

    let headerFound = false;
    let headerSize = 0;

    lines.forEach(line => {
        const parsedLine = parseCSVLine(line);

        if (!headerFound) {
            if (isRealHeader(parsedLine)) {
                data.headers = parsedLine;
                headerFound = true;
                headerSize = parsedLine.length;
            }
            return;
        }

        while (parsedLine.length < headerSize) {
            parsedLine.push("");
        }

        if (isRowEmpty(parsedLine)) {
            return;
        }

        data.rows.push(parsedLine);
    });

    if (!headerFound) {
        throw new Error("CSV header row was not found.");
    }

    return data;
}

// chooses only meaningful columns for display
function getDisplayColumns(headers) {
    const displayColumns = [];

    headers.forEach((header, index) => {
        const cleanHeader = header.trim();
        if (cleanHeader !== "" && cleanHeader.toLowerCase() !== "column1") {
            displayColumns.push(index);
        }
    });

    return displayColumns;
}

// computes proper width for each displayed column
function computeColumnWidths(data, displayColumns) {
    return displayColumns.map(colIndex => {
        let maxWidth = data.headers[colIndex].length;

        data.rows.forEach(row => {
            const value = safeGet(row, colIndex);
            if (value.length > maxWidth) {
                maxWidth = value.length;
            }
        });

        return Math.min(maxWidth, 35);
    });
}

// formats each cell for aligned output
function formatCell(text, width) {
    text = text || "";

    if (text.length > width) {
        text = width > 3 ? text.substring(0, width - 3) + "..." : text.substring(0, width);
    }

    return text.padEnd(width, " ");
}

function printSeparator(widths, rowNumberWidth) {
    let line = "-".repeat(rowNumberWidth) + "-+-";
    line += widths.map(width => "-".repeat(width)).join("-+-");
    console.log(line);
}

async function main() {
    try {
        console.log("Student: CANTAROS, RICH ANDREI");
        console.log("MP12 - Display Dataset in Formatted Table Output");

        const filePath = await askQuestion("Enter CSV dataset file path: ");
        const data = loadCSV(filePath);

        if (data.rows.length === 0) {
            console.log("No valid data rows were found in the dataset.");
            rl.close();
            return;
        }

        const displayColumns = getDisplayColumns(data.headers);

        if (displayColumns.length === 0) {
            console.log("No displayable columns were found.");
            rl.close();
            return;
        }

        const widths = computeColumnWidths(data, displayColumns);
        const rowNumberWidth = Math.max(4, String(data.rows.length).length + 2);

        console.log("\n========== FORMATTED DATASET TABLE ==========");
        console.log(`Total Valid Rows: ${data.rows.length}`);
        console.log("Displayed Columns: meaningful columns only");
        console.log("=============================================\n");

        printSeparator(widths, rowNumberWidth);

        let headerLine = formatCell("No.", rowNumberWidth) + " | ";
        headerLine += displayColumns
            .map((colIndex, i) => formatCell(data.headers[colIndex], widths[i]))
            .join(" | ");
        console.log(headerLine);

        printSeparator(widths, rowNumberWidth);

        data.rows.forEach((row, rowIndex) => {
            let rowLine = formatCell(String(rowIndex + 1), rowNumberWidth) + " | ";
            rowLine += displayColumns
                .map((colIndex, i) => formatCell(safeGet(row, colIndex), widths[i]))
                .join(" | ");
            console.log(rowLine);
        });

        printSeparator(widths, rowNumberWidth);

    } catch (error) {
        console.log("Error:", error.message);
    } finally {
        rl.close();
    }
}

main();